doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -128,45 +129,47 @@ Universe polymorphic rules
+will be detected if they also appear in a relevance mark in the pattern, such that
+they can be substituted when the rule is applied (otherwise you will get an undeclared sort quality error).
 
    .. coqtop:: reset all warn
 
       #[universes(polymorphic)] Symbol raise@{q|u|} : forall (A : Type@{q|u}), A.
       Rewrite Rule raise_nat :=
-        @{q|u+|+} |- raise@{q|u} (forall (x : ?A), ?P) => fun (x : ?A) => raise@{q|u} ?P.
+        raise@{q|u} (forall (x : ?A), ?P) => fun (x : ?A) => raise@{q|u} ?P.
 
 Rewrite rules, type preservation, confluence and termination
 ------------------------------------------------------------
 
-Currently, rewrite rules do not ensure that types must be preserved.
-There is a superficial check that the replacement needs to be typed
-against the type inferred for the pattern (for an unclear definition of type of a pattern),
-but it is known to be incomplete and only emits a warning if failed.
-This then means that reductions using rewrite rules have no reason to preserve well-typedness at all.
-The responsibility of ensuring type preservation falls on the user entirely.
+The rewrite rules are typechecked so that all substituted replacements
+have the type that the term being rewritten has (as described in :cite:`Rew24`).
+This means that the check is more stringent than ensuring the two sides have a shared type.
+This also means that the pattern is typechecked differently from regular terms,
+with an especially more restricted unification engine,
+which leads to the typechecker refusing rules that do respect the criterion.
+For this reason, the typechecker only emits warnings when it fails to verify the rule,
+letting the user take responsibility over the correctness of the rule.
 
-Similarly, neither confluence nor termination are checked by the compiler.
+Rocq currently doesn't check for confluence of the rewrite rules,
+even though it is required for the invariants that the typechecker uses.
+There are plans to add a check using the triangle criterion described in :cite:`TotR21`.
 
-There are future plans to add a check on confluence using the triangle criterion :cite:`TotR21`
-and a more complete check on type preservation.
+Rocq doesn't check for termination of the rewrite rules either.
+Indeed, non-terminating rules are generally fine except for one thing:
+the typechecker might not always terminate anymore.
+Unlike the previous two properties, non-termination cannot introduce crashes or anomalies.
 
 Compatibility with the eta laws
 -------------------------------
 
 Currently, pattern matching against rewrite rules pattern cannot do eta-expansion or contraction,
 which means that it cannot properly match against terms of functional types or primitive records.
-As with type preservation, a check is done to test whether this may happen,
-but it is not complete (false positives) and thus only emits a warning if failed.
+Rocq checks whether this may happen,
+but the check is not complete (false positives).
```

Comment:
```suggestion
but the check is imperfect (it reports false positives).
```

"(false positives)" alone is not clear.  I'm guessing this is what you mean.  "not complete" suggests some problems are not found when IIUC the problem is that reports false positives.